File: android/sdui-core/src/main/java/com/nba/sdui/core/data/AblyChannelManager.kt
package com.nba.sdui.core.data

import android.util.Log
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import io.ably.lib.realtime.AblyRealtime
import io.ably.lib.realtime.Channel
import io.ably.lib.realtime.ChannelState
import io.ably.lib.realtime.ConnectionState
import io.ably.lib.rest.Auth
import io.ably.lib.types.ClientOptions
import io.ably.lib.types.Message
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import java.net.HttpURLConnection
import java.net.URL

/**
 * Ably Channel Manager - Manages real-time connections to Ably for live data.
 *
 * Messages are returned as opaque `Map<String, Any?>` — the client has NO
 * knowledge of the payload structure.  The server's `dataBinding` on each
 * section define how individual fields map into section data.
 *
 * Implements:
 * - JWT token authentication via authUrl (NBA identity server)
 * - Channel subscription for real-time updates
 * - Flow-based message delivery to Compose UI
 */
class AblyChannelManager(
    private val tokenUrl: String
) {
    companion object {
        private const val TAG = "AblyChannelManager"
    }

    private val objectMapper = ObjectMapper().registerKotlinModule()

    private var ablyClient: AblyRealtime? = null
    private val activeChannels = mutableMapOf<String, Channel>()

    /** External listener for connection state changes (connected, disconnected, failed, etc.) */
    var onConnectionStateChange: ((ConnectionState) -> Unit)? = null

    /**
     * Initialize the Ably client with token authentication.
     *
     * Connection is non-blocking — the SDK connects in the background and
     * handles retries automatically. Channels can be subscribed immediately;
     * they attach once the connection is ready.
     *
     * Uses an authCallback instead of authUrl so we control the HTTP request
     * and can log the raw response for diagnostics.
     */
    fun initialize() {
        if (ablyClient != null) {
            Log.d(TAG, "Ably client already initialized")
            return
        }

        try {
            val options = ClientOptions().apply {
                autoConnect = true
                logLevel = io.ably.lib.util.Log.WARN

                authCallback = Auth.TokenCallback { _ ->
                    fetchTokenFromUrl(tokenUrl)
                }
            }

            ablyClient = AblyRealtime(options)

            ablyClient?.connection?.on { state ->
                Log.d(TAG, "Ably connection state: ${state.current}")
                onConnectionStateChange?.invoke(state.current)
                when (state.current) {
                    ConnectionState.connected -> Log.i(TAG, "Ably connected")
                    ConnectionState.disconnected -> Log.w(TAG, "Ably disconnected")
                    ConnectionState.failed -> Log.e(TAG, "Ably connection failed: ${state.reason}")
                    else -> {}
                }
            }

            Log.i(TAG, "Ably client created (tokenUrl: $tokenUrl)")

        } catch (e: Exception) {
            Log.e(TAG, "Failed to create Ably client", e)
            ablyClient = null
        }
    }

    /**
     * Fetch token from the auth URL and extract the JWT.
     *
     * The NBA identity endpoint wraps the token:
     *   {"status":"success","data":{"accessToken":"<JWT>"}}
     *
     * The Ably SDK expects the raw JWT string, not the wrapper.
     */
    private fun fetchTokenFromUrl(url: String): Any {
        val conn = URL(url).openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "GET"
            conn.connectTimeout = 10_000
            conn.readTimeout = 10_000
            conn.setRequestProperty("Accept", "application/json")

            val code = conn.responseCode
            val body = if (code in 200..299) {
                conn.inputStream.bufferedReader().readText()
            } else {
                val err = conn.errorStream?.bufferedReader()?.readText() ?: "(no body)"
                Log.e(TAG, "Token request failed: HTTP $code — $err")
                throw RuntimeException("Token endpoint returned HTTP $code")
            }

            Log.d(TAG, "Token response (HTTP $code, ${body.length} chars): ${body.take(200)}")

            val json = objectMapper.readTree(body)
            val jwt = json.path("data").path("accessToken").asText()
            if (jwt.isNotBlank()) {
                Log.d(TAG, "Extracted JWT (${jwt.length} chars)")
                return jwt
            }

            Log.d(TAG, "No wrapper detected, returning raw body to SDK")
            return body
        } finally {
            conn.disconnect()
        }
    }

    /**
     * Subscribe to a channel and receive messages as opaque Maps.
     *
     * The returned Flow emits the raw JSON payload as `Map<String, Any?>`.
     * Callers use [DataBindingResolver.applyBindings] with the section's
     * `dataBinding` to map fields into section data.
     */
    fun subscribeToChannel(channelName: String): Flow<Map<String, Any?>> = callbackFlow {
        Log.d(TAG, "Subscribing to channel: $channelName")

        val client = ablyClient ?: throw IllegalStateException("Ably client not initialized")

        val channel = client.channels.get(channelName)
        activeChannels[channelName] = channel

        channel.on { state ->
            Log.d(TAG, "Channel $channelName state: ${state.current}")
            when (state.current) {
                ChannelState.attached -> Log.i(TAG, "Channel attached: $channelName")
                ChannelState.detached -> Log.w(TAG, "Channel detached: $channelName")
                ChannelState.failed -> Log.e(TAG, "Channel failed: $channelName, reason=${state.reason}")
                else -> {}
            }
        }

        val messageListener = Channel.MessageListener { message: Message ->
            Log.d(TAG, "Received message on $channelName: ${message.name}")

            try {
                val parsed = parseMessage(message)
                if (parsed != null) {
                    trySend(parsed)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse message", e)
            }
        }

        channel.subscribe(messageListener)

        awaitClose {
            try {
                Log.d(TAG, "Unsubscribing from channel: $channelName")
                channel.unsubscribe(messageListener)
            } catch (e: Exception) {
                Log.e(TAG, "Error unsubscribing from channel: $channelName", e)
            } finally {
                activeChannels.remove(channelName)
            }
        }
    }

    /**
     * Disconnect and clean up.
     */
    fun disconnect() {
        Log.d(TAG, "Disconnecting Ably client")
        for ((name, channel) in activeChannels.toList()) {
            try {
                channel.unsubscribe()
            } catch (e: Exception) {
                Log.e(TAG, "Error unsubscribing channel during disconnect: $name", e)
            }
        }
        activeChannels.clear()
        try {
            ablyClient?.close()
        } catch (e: Exception) {
            Log.e(TAG, "Error closing Ably client", e)
        }
        ablyClient = null
    }

    /**
     * Parse an Ably message into an opaque Map.
     * No field-level knowledge of the payload — just raw JSON → Map.
     */
    @Suppress("UNCHECKED_CAST")
    private fun parseMessage(message: Message): Map<String, Any?>? {
        val data = message.data ?: return null

        return try {
            when (data) {
                is String -> objectMapper.readValue(data, Map::class.java) as Map<String, Any?>
                is Map<*, *> -> data as Map<String, Any?>
                else -> null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse message payload", e)
            null
        }
    }
}

File: ios/Sources/SduiCore/Realtime/AblyChannelManager.swift
import Foundation
import os

#if !SDUI_DISABLE_ABLY
import Ably
#endif

private let logger = Logger(subsystem: "com.nba.sdui", category: "AblyChannelManager")

#if SDUI_DISABLE_ABLY

// MARK: - No-op stub (SDUI_DISABLE_ABLY)
//
// Compiled when the package is built without the `ably-cocoa` dependency —
// e.g. on Intel Macs where Ably's SPM module map is broken on the x86_64
// simulator slice, or on any host that just wants polling-only runtime.
//
// The stub keeps the full public API so `SduiScreenViewModel` compiles
// unchanged. All calls are inert: subscriptions produce an empty
// `AsyncStream` that finishes immediately, and `connectionStates()`
// emits a single `.other` state before finishing so consumers that
// await a first event don't hang forever.

actor AblyChannelManager {

    enum ConnectionState: Sendable {
        case connected
        case disconnected
        case failed(message: String?)
        case other
    }

    init(tokenURL: URL) {
        logger.notice("AblyChannelManager stub active (SDUI_DISABLE_ABLY). Real-time via Ably is disabled; polling still works.")
    }

    func setChannelVisible(_ channelName: String, visible: Bool) {}

    func isChannelVisible(_ channelName: String) -> Bool { true }

    func initialize() {}

    func disconnect() {}

    func subscribe(channelName: String) -> AsyncStream<[String: Any]> {
        AsyncStream { $0.finish() }
    }

    func connectionStates() -> AsyncStream<ConnectionState> {
        AsyncStream { continuation in
            continuation.yield(.other)
            continuation.finish()
        }
    }
}

#else

/// Manages the SDUI app's real-time Ably connection. Mirrors Android's
/// [`AblyChannelManager`](../../android/sdui-core/src/main/java/com/nba/sdui/core/data/AblyChannelManager.kt).
///
/// Ably messages are opaque JSON — they surface as `[String: Any]`
/// dictionaries with no typed decoding. Callers route each payload through
/// ``DataBindingApplier`` to project it onto a section's current data.
///
/// Concurrency: `actor` isolation serialises channel registration /
/// teardown. Subscriptions expose `AsyncStream<[String: Any]>` so consumers
/// use plain `for await` loops inside `Task`s.
actor AblyChannelManager {

    enum ConnectionState: Sendable {
        case connected
        case disconnected
        case failed(message: String?)
        case other
    }

    private let tokenURL: URL
    private var client: ARTRealtime?
    private var channels: [String: ARTRealtimeChannel] = [:]
    private var subscriptions: [String: ARTEventListener] = [:]

    /// Channels marked invisible drop inbound messages rather than yielding
    /// them to consumers. Re-entering visibility starts yielding again.
    /// Missing entry = visible (default), so that the pre-visibility API
    /// remains backward-compatible for callers that don't bother reporting.
    private var channelVisibility: [String: Bool] = [:]

    /// Exposed as an AsyncStream for consumers that want to react to
    /// connection-state transitions (the VM flips `SectionStalenessTracker`
    /// on sustained disconnects).
    private var connectionContinuations: [UUID: AsyncStream<ConnectionState>.Continuation] = [:]

    init(tokenURL: URL) {
        self.tokenURL = tokenURL
    }

    /// Mark a channel visible / invisible. While invisible the manager
    /// reports `isChannelVisible == false`; consumers (the VM) decide
    /// whether to skip message processing. The actual Ably subscription
    /// stays attached to avoid the re-handshake cost when the section
    /// scrolls back into view.
    func setChannelVisible(_ channelName: String, visible: Bool) {
        channelVisibility[channelName] = visible
        logger.debug("channel \(channelName, privacy: .public) visible=\(visible)")
    }

    /// Query the current visibility state for a channel. Missing
    /// entries default to `true` (visible) so callers that never report
    /// visibility retain pre-Phase-F behaviour.
    func isChannelVisible(_ channelName: String) -> Bool {
        channelVisibility[channelName] ?? true
    }

    // MARK: - Lifecycle

    /// Idempotently create the underlying `ARTRealtime` client. Token
    /// retrieval happens lazily via the auth callback.
    func initialize() {
        guard client == nil else { return }

        let options = ARTClientOptions()
        options.autoConnect = true
        options.logLevel = .warn
        options.authCallback = { [tokenURL] _, callback in
            Task {
                await Self.fetchToken(from: tokenURL, callback: callback)
            }
        }

        let realtime = ARTRealtime(options: options)
        realtime.connection.on { [weak self] stateChange in
            guard let self else { return }
            let mapped: ConnectionState
            switch stateChange.current {
            case .connected: mapped = .connected
            case .disconnected, .suspended: mapped = .disconnected
            case .failed: mapped = .failed(message: stateChange.reason?.message)
            default: mapped = .other
            }
            Task { await self.broadcastConnection(mapped) }
        }
        client = realtime
        logger.info("Ably client created tokenURL=\(self.tokenURL.absoluteString, privacy: .public)")
    }

    func disconnect() {
        for (name, listener) in subscriptions {
            if let channel = channels[name] {
                channel.unsubscribe(listener)
            }
        }
        subscriptions.removeAll()
        channels.removeAll()
        client?.close()
        client = nil
        for (_, cont) in connectionContinuations {
            cont.finish()
        }
        connectionContinuations.removeAll()
    }

    // MARK: - Subscriptions

    /// Subscribe to a channel by name. Returns an `AsyncStream` that emits
    /// each message payload as an opaque `[String: Any]`.
    func subscribe(channelName: String) -> AsyncStream<[String: Any]> {
        initialize()
        guard let client else {
            return AsyncStream { $0.finish() }
        }

        let channel = client.channels.get(channelName)
        channels[channelName] = channel

        channel.on { stateChange in
            switch stateChange.current {
            case .attached: logger.info("channel attached \(channelName, privacy: .public)")
            case .detached: logger.warning("channel detached \(channelName, privacy: .public)")
            case .failed: logger.error("channel failed \(channelName, privacy: .public) reason=\(stateChange.reason?.message ?? "?", privacy: .public)")
            default: break
            }
        }

        return AsyncStream { continuation in
            let listener = channel.subscribe { [channelName] message in
                guard let payload = Self.parseMessage(message) else {
                    logger.warning("dropping non-dictionary payload on \(channelName, privacy: .public)")
                    return
                }
                continuation.yield(payload)
            }

            if let listener {
                Task { [weak self] in
                    await self?.registerListener(listener, for: channelName)
                }
            }

            continuation.onTermination = { [weak self] _ in
                Task { [weak self] in await self?.tearDown(channelName: channelName) }
            }
        }
    }

    /// Connection-state stream; a fresh stream per consumer (VM + tests).
    func connectionStates() -> AsyncStream<ConnectionState> {
        AsyncStream { continuation in
            let id = UUID()
            Task { [weak self] in
                await self?.registerConnectionContinuation(id: id, continuation: continuation)
            }
            continuation.onTermination = { [weak self] _ in
                Task { [weak self] in await self?.unregisterConnectionContinuation(id: id) }
            }
        }
    }

    // MARK: - Private

    private func registerListener(_ listener: ARTEventListener, for channelName: String) {
        subscriptions[channelName] = listener
    }

    private func tearDown(channelName: String) {
        if let listener = subscriptions.removeValue(forKey: channelName),
           let channel = channels[channelName] {
            channel.unsubscribe(listener)
        }
        channels.removeValue(forKey: channelName)
    }

    private func registerConnectionContinuation(id: UUID, continuation: AsyncStream<ConnectionState>.Continuation) {
        connectionContinuations[id] = continuation
    }

    private func unregisterConnectionContinuation(id: UUID) {
        connectionContinuations.removeValue(forKey: id)
    }

    private func broadcastConnection(_ state: ConnectionState) {
        for (_, cont) in connectionContinuations {
            cont.yield(state)
        }
    }

    private static func parseMessage(_ message: ARTMessage) -> [String: Any]? {
        switch message.data {
        case let dict as [String: Any]:
            return dict
        case let str as String:
            guard let data = str.data(using: .utf8) else { return nil }
            return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        case let data as Data:
            return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        default:
            return nil
        }
    }

    /// The NBA identity endpoint wraps tokens as
    /// `{"status":"success","data":{"accessToken":"<JWT>"}}` — Ably expects
    /// the raw JWT so we unwrap before handing it back.
    private static func fetchToken(
        from url: URL,
        callback: @escaping (ARTTokenDetailsCompatible?, (any Error)?) -> Void
    ) async {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
                let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
                throw NSError(domain: "AblyAuth", code: statusCode, userInfo: [
                    NSLocalizedDescriptionKey: "Token endpoint returned HTTP \(statusCode)"
                ])
            }
            let jwt = extractJWT(from: data) ?? String(data: data, encoding: .utf8) ?? ""
            callback(jwt as NSString, nil)
        } catch {
            logger.error("Ably token fetch failed: \(error.localizedDescription, privacy: .public)")
            callback(nil, error)
        }
    }

    private static func extractJWT(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let innerData = json["data"] as? [String: Any],
              let token = innerData["accessToken"] as? String,
              !token.isEmpty else { return nil }
        return token
    }
}

#endif
